### PR TITLE
[NEW] Update user profile

### LINKED ISSUE #39: [NEW] Support method to update profile

### PR DESCRIPTION
**Pull request by filipedelimabrito**
Closes #39 

- [x] Update email address for the user.
- [x] Update display name of the user.
- [x] Update password for the user.
- [x] Update username for the user.
- [x] Reset an user’s avatar.
- [x] Set an user’s avatar (from a file image).
- [x] Set an user’s avatar (from an URL).
- [x] Unit tests for the features above.

Documentation:
- https://rocket.chat/docs/developer-guides/rest-api/users/update/
- https://rocket.chat/docs/developer-guides/rest-api/users/resetavatar/
- https://rocket.chat/docs/developer-guides/rest-api/users/setavatar/

### CODE REVIEW

File: core/src/main/kotlin/chat/rocket/core/internal/model/UserUpdateEmailPayload.kt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package chat.rocket.core.internal.model
+
+import se.ansman.kotshi.JsonSerializable
+
+@JsonSerializable
+class UserUpdateEmailPayload(val userId: String, val email: String)
```
Comment (luciofm) [inline core/src/main/kotlin/chat/rocket/core/internal/model/UserUpdateEmailPayload.kt]: If the user updates email and name we will have to make 2 calls? Doesn't it better to aggregate all the user info in just one object? Send current info + edited fields.

Comment (filipedelimabrito) [reply core/src/main/kotlin/chat/rocket/core/internal/model/UserUpdateEmailPayload.kt]: I partly agree with you. The idea here is to make "specific calls" on the `rocketChatClient` from our mobile app. 

I'm taking into account the usage from the mobile app: we'll show to the user a view with its profile information. For instance, when the user touches the email field we'll show a new view allowing the user to edits the email. This way we'll have more specific calls (avoiding null parameters or sending the same value to the calls).

But if we plan to allow the users to edit their profile from a single view, aggregating all user info and calling a single method sounds better.

Here is an illustration from the Twitter and Instagram app just for reference:

photo_2018-01-18_09-04-49

photo_2018-01-18_09-04-39

photo_2018-01-18_09-04-44

photo_2018-01-18_09-04-47

**Comment by leonardoaramaki** [comment]
Maybe we could have the specific calls you mentioned and another one like Lucio said, with a payload containing all the data. That way we would contemplate all cases where the app the uses the ask wanted an app that updates the profile partially or when it wants to update everything at once at the end. What do you think?

**Comment by filipedelimabrito** [comment]
@leonardoaramaki My main concern was about the call sending null values. After debugging I saw that null proprieties is ignored on the `payloadBody` object. I'm making changes here... 
Thank you guys.
c/c @luciofm

**Comment by leonardoaramaki** [comment]
:+1:

File: core/src/main/kotlin/chat/rocket/core/internal/model/UserPayload.kt [hunk 2]
```diff
@@ -1,11 +1,8 @@
...
-
-@JsonSerializable
-data class UserPayload(
-        val email: String,
-        val name: String,
-        val pass: String,
-        val username: String
-)
\ No newline at end of file
+class UserPayload(val userId: String?,
```
Comment (luciofm) [inline core/src/main/kotlin/chat/rocket/core/internal/model/UserPayload.kt]: Isn't the userId always required?

Comment (filipedelimabrito) [reply core/src/main/kotlin/chat/rocket/core/internal/model/UserPayload.kt]: Not always. You can check it here: https://rocket.chat/docs/developer-guides/rest-api/users/setavatar/

Also, I made the test to ensure that everything works.

File: core/src/main/kotlin/chat/rocket/core/internal/rest/User.kt [hunk 3]
```diff
@@ -25,6 +33,99 @@ suspend fun RocketChatClient.me(): Myself {
...
+}
+
+/**
+ * Sets the user's avatar.
+ *
+ * @param avatarUrl Url of the avatar for the user
+ *
+ * @return True if the avatar was setted up, false otherwise.
+ */
+suspend fun RocketChatClient.setAvatar(avatarUrl: String): BaseResult {
```
[the remark this answers is not in the document]
Comment (filipedelimabrito) [reply core/src/main/kotlin/chat/rocket/core/internal/rest/User.kt]: 👍

File: core/src/main/kotlin/chat/rocket/core/internal/model/UserPayload.kt (hunk 2 above)
Comment (luciofm) [reply core/src/main/kotlin/chat/rocket/core/internal/model/UserPayload.kt]: But this is another call, they should use different objects.

**Pull request merged**